File: 142.linked-list-cycle-ii.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func detectCycle(head *ListNode) *ListNode {
	nodeAddrs := make(map[*ListNode]bool)
	p := head
	for p != nil {
		if _, ok := nodeAddrs[p]; ok {
			return p
		}
		nodeAddrs[p] = true
		p = p.Next
	}
	return nil
}

func detectCycle2(head *ListNode) *ListNode {
	// Using Floyd Detection
	if head == nil {
		return nil
	}
	p1, p2 := head, head
	if p1.Next == nil || p2.Next == nil || p2.Next.Next == nil {
		return nil
	}
	p1 = p1.Next
	p2 = p2.Next.Next
	for *p1 != *p2 {
		if p1.Next == nil || p2.Next == nil || p2.Next.Next == nil {
			return nil
		}
		p1 = p1.Next
		p2 = p2.Next.Next
	}
	p1 = head
	for *p1 != *p2 {
		p1 = p1.Next
		p2 = p2.Next
	}
	return p1
}

File: 142.linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None or head.next.next is None:
            return None
        slow = head.next
        fast = head.next.next
        while slow is not None and fast is not None and fast.next is not None:
            if slow == fast:
                s1 = slow
                h1 = head
                while s1 != h1:
                    s1 = s1.next
                    h1 = h1.next
                return s1
            slow = slow.next
            fast = fast.next.next
        return None